pkg/standalone-utils/contracts/BalancerSorQueries.sol

```diff
@@ -0,0 +1,293 @@
+        returns (
+            uint256[][] memory balances,
+            uint256[] memory totalSupplies,
+            uint256[] memory swapFees,
+            uint256[] memory linearWrappedTokenRates,
+            uint256[][] memory weights,
+            uint256[][] memory scalingFactors,
+            uint256[] memory amps
+        )
+    {
+        uint256 i;
+        address[] memory pools = new address[](poolIds.length);
+
+        for (i = 0; i < poolIds.length; i++) {
+            (pools[i], ) = vault.getPool(poolIds[i]);
+        }
+
+        if (config.loadTokenBalanceUpdatesAfterBlock) {
+            balances = getPoolTokenBalancesWithUpdatesAfterBlock(poolIds, config.blockNumber);
+        }
+
+        if (config.loadTotalSupply) {
+            totalSupplies = getTotalSupplyForPools(pools, config.totalSupplyTypes);
+        }
+
+        if (config.loadSwapFees) {
+            swapFees = getSwapFeePercentageForPools(pools, config.swapFeeTypes);
+        }
+
+        if (config.loadLinearWrappedTokenRates) {
+            address[] memory linearPools = new address[](config.linearPoolIdxs.length);
+
+            for (i = 0; i < config.linearPoolIdxs.length; i++) {
+                linearPools[i] = pools[config.linearPoolIdxs[i]];
+            }
+
+            linearWrappedTokenRates = getWrappedTokenRateForLinearPools(linearPools);
+        }
+
+        if (config.loadNormalizedWeights) {
+            address[] memory weightedPools = new address[](config.weightedPoolIdxs.length);
+
+            for (i = 0; i < config.weightedPoolIdxs.length; i++) {
+                weightedPools[i] = pools[config.weightedPoolIdxs[i]];
+            }
+
+            weights = getNormalizedWeightsForPools(weightedPools);
+        }
+
+        if (config.loadScalingFactors) {
+            address[] memory scalingFactorPools = new address[](config.scalingFactorPoolIdxs.length);
+
+            for (i = 0; i < config.scalingFactorPoolIdxs.length; i++) {
+                scalingFactorPools[i] = pools[config.scalingFactorPoolIdxs[i]];
+            }
+
+            scalingFactors = getScalingFactorsForPools(scalingFactorPools);
+        }
+
+        if (config.loadAmps) {
+            address[] memory ampPools = new address[](config.ampPoolIdxs.length);
+
+            for (i = 0; i < config.ampPoolIdxs.length; i++) {
+                ampPools[i] = pools[config.ampPoolIdxs[i]];
+            }
+
+            amps = getAmpForPools(ampPools);
+        }
+    }
+
+    function getPoolTokenBalancesWithUpdatesAfterBlock(bytes32[] memory poolIds, uint256 blockNumber)
+        public
+        view
+        returns (uint256[][] memory)
+    {
+        uint256[] memory balances;
+        uint256 lastChangeBlock;
+        uint256[][] memory allBalances = new uint256[][](poolIds.length);
+
+        for (uint256 i = 0; i < poolIds.length; i++) {
+            (, balances, lastChangeBlock) = vault.getPoolTokens(poolIds[i]);
+
+            if (lastChangeBlock > blockNumber) {
```

Comment:
I'm not sure this logic makes sense and can't think of scenarios where this is true. Since the `getPoolTokens` call happens no matter what, why not always return all balances since that is assumed to be the latest state